### ISSUE TITLE
Carla agent performance and training time

### ISSUE THREAD

**Issue by LocLePhuoc** [PacktPublishing/Hands-On-Intelligent-Agents-with-OpenAI-Gym#25]
@praveen-palanisamy thank you for your extremely helpful code base. I have some questions that I hope you could give some insights into:

- I noticed that the reward function is the one that was introduced in the 2017 Carla paper. The team who wrote the paper trained A2C with 10 agents for 10 million steps and they noted that even then their final agent is not a good one. Were you able to achieve a reasonably good agent, and in how long of training ?

- I tried reimplementing non async A2C with the same reward function and discrete action space because my limited hardware can't handle multiple agents. Can I achieve the same level of performance to that of the async version of A2C ? So far, after several hundred thousand steps, the agent seems to only move straight and cannot make any turn at all, is that an expected behavior ?   

-  In the scenario file, I noticed that you only use one pose in each scenario, did you train on only one too. Currently my implementation is switching back and forth between poses in the list of straight poses, curve poses etc. Will that make any difference ?

Once again, thank you for your time and insights.

**Comment by praveen-palanisamy** [comment]
Hey @LocLePhuoc : Glad that the code base is useful to you.

  - The A2C agent in this repository trained for about 4M steps gave some reasonable results (driving through turns, intersections and avoiding some obstacles). The original paper trained using ~ 1B frames/steps using asynchronous agents (A**3**C)

 - You could used batched version of Advantage Actor Critic if you can't run multiple actors in parallel due to hardware limitations. You can achieve similar (sometimes better is also possible) results but you will have to train for similar amount of steps which may take a little longer (in terms of wall-clock time) compared to asynchronous training methods. There are other training methods that are more sample/wall-clock-time efficient (eg.: Apex, R2D2) but they do require a lot of processing cores (hardware intensive).
 If the agent is able to keep-lanes/drive-straight,  make sure that the environment you train the agent has turns so that the agent learns to turn as well. After 600k steps or so, the agent should is usually able to make turns well. 

  - The default example code only trains on the straight poses to make the task easy and therefore lesser training time for the users. You are headed in the correct direction for improving the driving performance of the agent. Having a mix of straight, curve and nav (obstacles) poses will increase the variance in the input data for the agent and as the agent trained in such an environment, it will be able to better adapt/generalize to the test driving scenarios.

Hope that helps.

**Comment by LocLePhuoc** [comment]
Thank you @praveen-palanisamy for your advices.

I'm currently using a pretrained model for image feature extraction, do you think it will impact the training in any negative way ??

So far I've trained my agent to learn how to turn in one pose only. But the agent seems to turn to soon and ends up crashing into poles or sidewalks. The current number of steps is 300k - ish.

**Comment by praveen-palanisamy** [comment]
If the pre-trained feature extractor is not part of agent's policy, the agent's observation representation will be limited to what the feature extractor provides rather than the actual observation from the environment. So, the policy will be very dependent on the pre-trained feature extractor you are using. The agent's policy will probably not work well if you updated the weights of the feature extractor or used a different perception head.
But that said, if the feature extractor is doing a good job of reducing the high dimensional RGB/camera images to a low-dimensional yet feature rich space, it actually can aid and speedup the policy learning process.

In majority of my experiments, I use a policy that includes a few perception/feature extraction layers in itself. 

You mentioned that you are using the Carla-v0 with discrete action space. For high-dimensional, pixel-to-control type problems, even for discrete action spaces, the number of samples/steps required are usually large. The Atari agents that do pixel-to-control and reach human level performance usually need about 50M to 1B frames. The Carla-v0 driving environment is relatively much harder. So, don't loose hope in your agent! It just needs more samples :)

**Comment by LocLePhuoc** [comment]
I see, from what I can tell, the agent needs a lot of frames to just be able to eliminate unnecessary actions  (in the case of one curve only). I also noted that the longer the trajectory, the harder it is for the agent to actually find the right actions throughout.

Thank you so much for these insights, I will keep training them to see if the agent can make a turn consistently =))
